Calculate publisher confirm latency

acogoluegnes opened the pull request:
[#162152758]

Fixes #121

acogoluegnes commented:
OK, applied change as @michaelklishin suggested:

Publisher confirm off:
```
id: test-142837-410, time: 1.000s, sent: 900 msg/s, received: 899 msg/s, min/median/75th/95th/99th message latency: 129/427/856/3108/4744 µs
```

Publisher confirm on:
```
id: test-142636-522, time: 2.009s, sent: 11080 msg/s, confirmed: 11080 msg/s, nacked: 0 msg/s, received: 10897 msg/s, min/median/75th/95th/99th message latency: 42936/84395/103139/125290/150685 µs, confirm latency: , 44024/82644/101174/131103/155386 µs
```

michaelklishin commented:
@kjnilsson @gerhard does this look like what you folks had in mind?

gerhard commented:
This looks great!

Can we think of a better name for `message latency`? `confirm latency` is spot-on, I'm proposing `consume latency` or `end-to-end latency` instead of `message latency`. WDYT?

michaelklishin commented:
Roundtrip latency?

On Sat, Nov 24, 2018 at 12:36 AM Gerhard Lazu <notifications@github.com>
wrote:

> This looks great!
>
> Can we think of a better name for message latency? confirm latency is
> spot-on, I'm proposing consume latency or end-to-end latency instead of message
> latency. WDYT?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-perf-test/pull/122#issuecomment-441282617>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAEQoUEtFBW8SbqfQG787GYSD3gCcWhks5uyCQjgaJpZM4YwXk7>
> .
>


-- 
MK

Staff Software Engineer, Pivotal/RabbitMQ

michaelklishin commented:
Actually it's only roundtrip if a single PerfTest node runs both publishers and consumers. So maybe "message latency" is not too bad then.

gerhard commented:
What is `message latency` measuring when the PerfTest instance runs just a publisher? What if it runs just a consumer? And when the same instance runs both a publisher & consumer?

michaelklishin commented:
@gerhard when there are no consumers the metric is not reported. Which led me to another discovery: with `-y 0 -x 1 --confirms 100` this branch does not report confirm latency metrics. @acogoluegnes should it?

michaelklishin commented:
Inspecting the code suggests that the message latency histogram is only updated from `Consumer#handleDelivery` so renaming it to "consumer latency" sounds like a good idea.

acogoluegnes commented:
@michaelklishin @gerhard Thanks for taking a look. I renamed `message latency` to `consumer latency`. For `-y 0 -x 1 --confirms 100`, the confirm latency is now reported on the console. I refactored the class responsible for the console output and added some tests for it, as historically it had none.

gerhard commented:
Wording looks right, but sent rate is 10x off. perf-test reports `106430 msg/s`, rabbitmq-management reports ~10k msg/s:

```
~ docker run -it --rm -p 15672:15672 -p 5672:5672 rabbitmq:3.7.8-management
```
![image](https://user-images.githubusercontent.com/3342/49240973-7251b400-f3fe-11e8-869d-63b6bf9b1e94.png)

```
~/github.com/rabbitmq/rabbitmq-perf-test make run
[INFO] Scanning for projects...
[INFO]
[INFO] -----------------------< com.rabbitmq:perf-test >-----------------------
[INFO] Building RabbitMQ Performance Testing Tool 2.4.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- buildnumber-maven-plugin:1.4:create (default) @ perf-test ---
[INFO] Executing: /bin/sh -c cd '/Users/gerhard/github.com/rabbitmq/rabbitmq-perf-test' && 'git' 'rev-parse' '--verify' 'HEAD'
[INFO] Working directory: /Users/gerhard/github.com/rabbitmq/rabbitmq-perf-test
[INFO] Storing buildNumber: 03cce8ee7366214ae1881f49ee7e105245af8e30 at timestamp: 1543513210754
[INFO] Storing buildScmBranch: rabbitmq-perf-test-121-publish-confirm-latencies
[INFO]
[INFO] --- build-helper-maven-plugin:3.0.0:add-source (add-source) @ perf-test ---
[INFO] Source directory: /Users/gerhard/github.com/rabbitmq/rabbitmq-perf-test/src/java-packaging/java added.
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ perf-test ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 5 resources
[INFO] Copying 5 resources to static
[INFO]
[INFO] --- maven-compiler-plugin:3.8.0:compile (default-compile) @ perf-test ---
[INFO] Nothing to compile - all classes are up to date
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.097 s
[INFO] Finished at: 2018-11-29T17:40:11Z
[INFO] ------------------------------------------------------------------------
[INFO] Scanning for projects...
[INFO]
[INFO] -----------------------< com.rabbitmq:perf-test >-----------------------
[INFO] Building RabbitMQ Performance Testing Tool 2.4.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- exec-maven-plugin:1.6.0:java (default-cli) @ perf-test ---
id: test-174012-798, starting consumer #0
id: test-174012-798, starting consumer #0, channel #0
id: test-174012-798, starting producer #0
id: test-174012-798, starting producer #0, channel #0
id: test-174012-798, time: 1.000s, sent: 80470 msg/s, received: 9575 msg/s, min/median/75th/95th/99th consumer latency: 17689/366301/551722/676050/682473 µs
id: test-174012-798, time: 2.000s, sent: 106430 msg/s, received: 11168 msg/s, min/median/75th/95th/99th consumer latency: 707006/1076516/1293066/1513904/1537798 µs
id: test-174012-798, time: 3.000s, sent: 110971 msg/s, received: 10484 msg/s, min/median/75th/95th/99th consumer latency: 1546515/1981460/2204798/2385291/2414357 µs
...
```

acogoluegnes commented:
@gerhard Does this behavior persist or does it go away after a few seconds? I can observe the same thing on my laptop during the first seconds of the run, then the situation stabilizes, and what PerfTest and the broker report is the same, give or take.

I think PerfTest hammers the broker during the very first seconds and then flow control kicks in, the producer is throttled.

The rate calculation hasn't changed in this PR and all the rates are calculated with the same code, I'd be surprised this PR broke existing stats so badly. Note I observe the same behavior with PerfTest 2.3.0.

gerhard commented:
This has nothing to do with PerfTest, it's how I was running it.

TL;DR **Do not run PerfTest on macOS against RabbitMQ running in Docker Desktop**

PerfTest runs on the host, macOS v10.14.1 & RabbitMQ v3.7.8 runs inside Docker Desktop CE 2.0.0.0-mac81. Even though the PerfTest publisher is running at 120k-130k msg/s, most of these messages get buffered at the networking layer. I can see Docker memory usage growing from 2GB to 8GB within minutes. I can also observe consumer latency constantly increasing since the  consumer is limited to ~10k msg/s due to the Docker networking implementation in macOS. As soon as Docker hits the memory limit set in the Docker Desktop app - 8GB in my case - the networking stack grinds to a halt: existing TCP connections get terminated abruptly, new ones cannot be established. What's worse, the network buffer doesn't get cleared until Docker gets restarted. 🤦‍♂️

gerhard commented:
PR looking good, happy to see this merged.

I started looking into Docker Compose so that PerfTest dev builds can run against RabbitMQ with a single command, without polluting the host - everything runs in a container, with a single command: `docker compose`.

acogoluegnes commented:
@gerhard Could not reproduce on my MBP.

```
uname -a
Darwin MacBook-Pro-4.local 15.6.0 Darwin Kernel Version 15.6.0: Mon Nov 13 21:58:35 PST 2017; root:xnu-3248.72.11~1/RELEASE_X86_64 x86_64
docker --version
Docker version 18.03.1-ce, build 9ee9f40
```

gerhard commented:
I can reproduce it consistently on mine:

```
uname -a
Darwin focker 18.2.0 Darwin Kernel Version 18.2.0: Fri Oct  5 19:41:49 PDT 2018; root:xnu-4903.221.2~2/RELEASE_X86_64 x86_64
docker --version
Docker version 18.09.0, build 4d60db4
```

It's a combination of many things that make this possible, but mostly a crazy fast host & a poor implementation of Docker Desktop. Not to worry though, a docker-compose-based workflow will make this irrelevant.
